pkg/ir/deps/sql_program_deps_graph.go

```diff
@@ -0,0 +1,256 @@
+
+	for idx, stmt := range parsedProgram {
+		inputs := []*Table{}
+		outputs := []*Table{}
+		for _, t := range stmt.Inputs {
+			fullName := "table." + t
+			tableNodelist, ok := tableNodeMap[fullName]
+			if !ok {
+				tableNode := &Table{
+					Type:        TypeTable,
+					Name:        t,
+					HazardIndex: 0,
+				}
+				tableNodeMap[fullName] = []*Table{tableNode}
+				inputs = append(inputs, tableNode)
+			} else {
+				inputs = append(inputs, tableNodelist[len(tableNodelist)-1])
+			}
+		}
+		for _, t := range stmt.Outputs {
+			fullName := "table." + t
+			tableNodeList, ok := tableNodeMap[fullName]
+			if !ok {
+				tableNode := &Table{
+					Type:        TypeTable,
+					Name:        t,
+					HazardIndex: 0,
+				}
+				tableNodeMap[fullName] = []*Table{tableNode}
+				outputs = append(outputs, tableNode)
+				continue
+			}
+			// find last statement that read/write this table
+			for j := len(result) - 1; j >= 0; j-- {
+				prev := result[j]
+				tableNodeList = tableNodeMap[fullName]
+				if contains(prev.Outputs, fullName) {
+					fmt.Printf("prev append WAW: %s, table: %s\n", prev.Statement, fullName)
+					// WAW solution
+					prevOutput, ok := getFirstInList(prev.Outputs, fullName)
+					if !ok {
+						return nil, fmt.Errorf("error adding WAW dependency, tablename: %s, prev stmt: %s", fullName, prev.Statement)
+					}
+					inputs = append(inputs, prevOutput)
+				} else if contains(prev.Inputs, fullName) {
+					fmt.Printf("prev append WAR: %s, table: %s\n", prev.Statement, fullName)
+					// WAR solution
+					readAsOutputTable := &Table{
+						Type:        TypeTable,
+						Name:        t,
+						HazardIndex: tableNodeList[len(tableNodeList)-1].HazardIndex + 1,
+					}
+					prev.Outputs = append(prev.Outputs, readAsOutputTable)
+					if !contains(inputs, fullName) {
+						inputs = append(inputs, readAsOutputTable)
+					}
+					tableNodeMap[fullName] = append(tableNodeMap[fullName], readAsOutputTable)
+				}
+			}
+			tableNodeList = tableNodeMap[fullName]
+			warTable := &Table{
+				Type:        TypeTable,
+				Name:        t,
+				HazardIndex: tableNodeList[len(tableNodeList)-1].HazardIndex + 1,
+			}
+			outputs = append(outputs, warTable)
+			tableNodeMap[fullName] = append(tableNodeMap[fullName], warTable)
+		}
+		curr := &Statement{
+			Statement: stmt.Original,
+			Order:     idx,
+			Inputs:    inputs,
+			Outputs:   outputs,
+		}
+		result = append(result, curr)
+	}
+	if err := drawGraphviz(result); err != nil {
+		return result, err
+	}
+	return result, nil
+}
+
+func getFirstInList(tableList []*Table, tableNameToFind string) (*Table, bool) {
+	for _, t := range tableList {
+		if t.FullName() == tableNameToFind {
+			return t, true
+		}
+	}
+	return nil, false
+}
+
+func contains(tableList []*Table, item string) bool {
+	res := false
+	for _, t := range tableList {
+		if t.FullName() == item {
+			res = true
+			break
+		}
+	}
+	return res
+}
+
+func drawGraphviz(stmts []*Statement) error {
```

Comment:
This function definition is too long. I am afraid that it would bring down our gocyclo96 score in our Go Score Card.